datafusion/core/src/catalog_common/information_schema.rs

```diff
@@ -406,6 +406,7 @@ fn get_udf_args_and_return_types(
             .into_iter()
             .map(|arg_types| {
                 // only handle the function which implemented [`ScalarUDFImpl::return_type`] method
+                #[allow(deprecated)]
```

Comment:
cc @goldmedal for ideas how (if) we will be able to resolve this deprecation.

Reply:
Thanks for reminding this. I filed https://github.com/apache/datafusion/issues/13735 to trace it.

Reply:
While reading https://github.com/apache/datafusion/issues/13735 and thinking about the churn we had recently with `invoke_with_batch` and `invoke_with_args`, etc

What if we made a new API that could accomodate both usecases. Something like

```rust
struct ReturnTypeArgs {
       // Argments, if available. This may not be available in some contexts
       // such as information_table schema queries
        args: Option<&[Expr]>
       // schema, if available. 
        schema: Option<&dyn ExprSchema>,
        arg_types: &[DataType],
}

impl ScalarUdfImpl {
  /// returns the result type, given `args` if possible
  /// if not possible, returns `Ok(None)`
  fn return_type_with_args(&self, args: &ReturnTypeArgs) -> Result<Option<DataType>> 
}
```

🤔

Reply:
can we change the signature to `return_type(ReturnTypeArgs)` and remove `return_type_with_args` at all

I was thinking about this too but not sure this huge breaking change is acceptable or not, but I think this will be a better change in long term

Reply:
But I guess for Comet, they need to switch to `return_type_with_args` first for their fix and then switch back to `return_type` after a while. Do you know who else rely on the `return_type` signature a lot and whether this breaking change is a pain for them?

Reply:
> But I guess for Comet, they need to switch to return_type_with_args first for their fix and then switch back to return_type after a while.

Yes that is true and would be non ideal I agree

> Do you know who else rely on the return_type signature a lot and whether this breaking change is a pain for them?

We use it a bunch in InfluxDB, eg https://github.com/search?q=repo%3Ainfluxdata%2Finfluxdb3_core%20return_type&type=code

I expect anyone who has created user defined functions has implemented `return_type` and thus will be impacted

Reply:
I like the idea of 'queuing up' any more api changes to udf's as much as possible (using deprecation that doesn't break any existing deployments) and release them all at once in 6 months or so. This includes the above, #13519, etc

Reply:
Yeah -- it will take more effort on the DataFusion side, but I think it will be much better on the downstream user side.

Now that DataFusion has more users, I think it makes sense that the balance will be pointing a bit back

Reply:
Let's go with `return_type_with_args` and deprecate the other two.

Reply:
> ```rust
> struct ReturnTypeArgs {
>        // Argments, if available. This may not be available in some contexts
>        // such as information_table schema queries
>         args: Option<&[Expr]>
>        // schema, if available. 
>         schema: Option<&dyn ExprSchema>,
>         arg_types: &[DataType],
> }
> ```



the seemingly single method `udf.return_type_with_args`  will be invokable with `ReturnTypeArgs` argument, which may or may not carry expressions. If the answer depends on the expression presence, it can depend in two ways

1. type is known when expressions are known (and are something particular), type is unknown otherwise
2. type is A when expressions are known (and are something particular), type is B (!= A) otherwise


The second result is clearly not acceptable from planning perspective (https://github.com/apache/datafusion/pull/13717#discussion_r1879725758). Retrieving a false type B is misleading. Type B cannot be used if the truth is type A. Which leads us to the conclusion that the first option is the only acceptable option.

For the first option, we realize that for some functions -- like for some syntaxes -- type is known when inputs are known. This is normal in query engines. This is normally handled by analyzer piece of a query engine, nothing special about it. This means that from query planning perspective, "what is this UDF return type" is a perfectly valid question with a perfectly well defined answer, that analyzer needs to know, should capture and persist. Once answered, there is no reason to ask this question ever again.

The remaining use-case seems to be `information_schema.routines` perspective. There, the return type is unknown (or is a set of possible types), because the callsite is undefined. 

I am therefore not convinced the introduction of `return_type_with_args` is a good approach. It just hides the problem, without addressing it. It's better if the problem is clearly visible. IMO the way to address the problem is what outlined above -- analyzer needs to know the type, so it should capture and persist that information -- https://github.com/apache/datafusion/issues/12604.

I believe we do not need to spend time on `return_type_with_args`, since it does not solve the crux of the problem.

Reply:
Let's turn the question around. What problems would `return_type_with_args` be solving?

Reply:
It is used to define the return type of the function, since they can have different return type based on the input (even it is coerced)

Reply:
It looks like this goal is already supported by the existing `return_type_from_exprs`.
What problem would `return_type_with_args` solve?

Reply:
We want single return_type for cases in #13735 that only takes `arg_types` and compatible with other cases that takes 3 args, so we switch to `return_type_with_args` and deprecate `return_type_from_exprs ` and `return_type`. Any further changing of field won't cause signature change as well.

Reply:
> We want single return_type for cases in #13735 that only takes `arg_types`

for information_schema we can solve this separately (eg udf could tell us what a set of possible return types is)

> and compatible with other cases that takes 3 args

The point is -- it's not compatible.
It's compatible in Rust compiler sense, but the Option in the struct means some information may or may not be present and the implementor may or may not use it. It's unnecessarily complex.

Reply:
The optionality is the same as the current status where you can use `return_type` for simple case or `return_type_from_exprs` for advanced case. 😕

Reply:
> The optionality is the same as the current status where you can use `return_type` for simple case or `return_type_from_exprs` for advanced case. 


Correct. And `return_type_with_args` would not solve this dichotomy problem.
It would only make it less clear and harder to implement correctly.

Since we're really solving information_schema problem now, we should think what could be the best solution for information_schema. **For example**, we could rename `return_type` or replace with `get_possible_return_types`, etc.

Reply:
> It would only make it less clear and harder to implement correctly.

Aside from the information_schema issue, keeping both `return_type` and `return_type_from_exprs` is more confusing than having a single return_type_with_args. Consolidating them into one does not worsen the current situation

Reply:
I think deprecate `return_type` and switch to `return_type_from_exprs` force the user to have `args` and `schema` which is not ideal, since they are optional parameter for certain function only. Without a doubt, they might pass the empty `vec![]` and empty schema ... which doesn't make sense. That is why `return_type_with_args` with optional parameter is a better choice.

Reply:
The information_schema seems like another independent problem. We can even just list the possible return types for it, because it doesn't care the decision of the return type, it cares about the `return_type` itself only. The trivial solution is introduce another trait implementation like you said.

`pub fn possible_return_types(&self) -> Vec<DataType>;`

Reply:
> Let's turn the question around. What problems would return_type_with_args be solving?

In my mind there are two usecases for `return_type`:
1. LogicalPlanning (aka the analyzer and type coercion, etc) when you have `Exprs` 
2. PhysicalPlanning  (aka physical optimizers)

As I understood the problem is that some things in the LogicalPlanner do things with the exprs (like determining the type of the arguments)

I agree with the "it is not clear when you get arguments in what contexts"

Maybe we could make this explicit with an enum like the following 🤔 

```rust
/// Information about arguments when calling `return_type`
enum ArgumentInfo {
  LogicalPlanning {
        args: &[Expr],
        schema: &dyn ExprSchema
   },
  PhysicalExecution, 
...
   
}

struct ReturnTypeArgs {
       // Argument information
       arg_info: ArgumentInfo,
       // argument DataTypes
        arg_types: &[DataType],
}
```

Reply:
> 2\. PhysicalPlanning  (aka physical optimizers)

For physical planning we should simply do equivalent of https://github.com/apache/datafusion/issues/13825 (but for physical exprs). Otherwise there is no good way to find out the return type. And complicating the call signature will only add insult to injury

Reply:
I think one of the challenge of #13825 and #12604 is that the properties of `Expr` including data type is determined based on `schema`. Unless the `schema` is the same all the way of the processing otherwise we need to recompute the properties based on given `schema` so we can't compute once and store the information within `Expr`. Introduce `Map` for schema -> properties mapping seems not practically.

Reply:
For example, if we have column `c`, we need `schema` to know it's data type is `Int32`. With another `schema`, it may becomes `Int64`. How do we ensure the schema is not changed and we can happily reuse the result we had before?